File: 0069-sqrtx/0069-sqrtx.rs
impl Solution {
    pub fn my_sqrt(x: i32) -> i32 {
        if x < 2 {
            return x;
        }

        let initial_guess = x / 2;

        let mut guess = initial_guess;
        let mut next_guess = (guess + x / guess) / 2;
        for _ in 0..100 {
            guess = next_guess;
            next_guess = (guess + x / guess) / 2;
        }
        if next_guess * next_guess > x {
            next_guess - 1
        } else {
            next_guess
        }
    }
}

File: 0069-sqrtx/0069-sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        low = 1
        high = x // 2
        ans = 1

        if x < 2:
            return x

        while low <= high:
            mid = low + ((high - low) // 2)
            pow_mid = mid * mid
            if pow_mid == x:
                return mid
            elif pow_mid < x:
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
        
        return ans